File: 2465-shifting-letters-ii/2465-shifting-letters-ii.py
class Solution:
    def shiftingLetters(self, s: str, shifts: List[List[int]]) -> str:
        s=list(s)
        toDo=[0]*(len(s)+1)
        for start,end,action in shifts:
            if action==0:
                toDo[start]-=1
                toDo[end+1]+=1
            else:
                toDo[start]+=1
                toDo[end+1]-=1
        shift=0
        for i in range(len(toDo)):
            shift+=toDo[i]
            toDo[i]=shift
        #print(toDo)
        for i in range(len(s)):
            s[i]=chr((ord(s[i])-ord('a')+toDo[i])%26+ord('a'))
        #print(s)
        return ''.join(s)

File: 2465-shifting-letters-ii/2465-shifting-letters-ii.java
class Solution {
    public String shiftingLetters(String str, int[][] shifts) {
        int l=str.length();
        char[] s=str.toCharArray();
        int[] toDo=new int[l+1];
        for(int i=0;i<shifts.length;i++){
            int start=shifts[i][0];
            int end=shifts[i][1];
            int action=shifts[i][2];
            if(action==0){
                toDo[start]-=1;
                toDo[end+1]+=1;
            }else{
                toDo[start]+=1;
                toDo[end+1]-=1;
            }
        }
        int shift=0;
        for(int i=0;i<toDo.length;i++){
            shift+=toDo[i];
            toDo[i]=shift;
        }
        StringBuilder result=new StringBuilder();
        for(int i=0;i<l;i++){
            int nc=((s[i]-'a'+toDo[i])%26);
            if(nc<0)nc+=26;
            result.append((char)(nc+'a'));
        }
        return result.toString();
    }
}